source/simulator/solver_schemes.cc

```diff
@@ -1452,6 +1455,155 @@ namespace aspect
+        assemblers->stokes_system_assembler_on_boundary_face_properties.need_viscosity = true;
+
+        assemblers->stokes_system_on_boundary_face.push_back(
+          std_cxx14::make_unique<aspect::Assemblers::StokesAdjointRHS<dim> >());
+
+        if (SimulatorAccess<dim> *p = dynamic_cast<SimulatorAccess<dim>* >(assemblers->stokes_system_on_boundary_face[0].get()))
+          p->initialize_simulator(*this);
+
+
+        // do pressure rhs compatibility
+        if (do_pressure_rhs_compatibility_modification)
+          assemblers->stokes_system.push_back(
+            std_cxx14::make_unique<aspect::Assemblers::StokesPressureRHSCompatibilityModification<dim> >());
+
+        if (SimulatorAccess<dim> *p = dynamic_cast<SimulatorAccess<dim>* >(assemblers->stokes_system[0].get()))
+          p->initialize_simulator(*this);
+
+
+        // Don't need to reassmble the stokes system because it's the same as for the forward problem
+        rebuild_stokes_matrix = rebuild_stokes_preconditioner = false;
+
+        // Assemble and solve the adjoint stokes system
+        assemble_stokes_system();
+        solve_stokes();
+
+        // save the solution in 'current_adjoint_solution'. This will be accessed during visualization and
+        // to calculate the sensitivity kernels
+        current_adjoint_solution.block(introspection.block_indices.velocities)
+          = solution.block(introspection.block_indices.velocities);
+        if (introspection.block_indices.velocities != introspection.block_indices.pressure)
+          current_adjoint_solution.block(introspection.block_indices.pressure)
+            = solution.block(introspection.block_indices.pressure);
+
+        // put forward solution back into solution vector so that the forward equation is processed during postprocessing
+        solution.block(introspection.block_indices.velocities) =
+          current_linearization_point.block(introspection.block_indices.velocities);
+        if (introspection.block_indices.velocities != introspection.block_indices.pressure)
+          solution.block(introspection.block_indices.pressure)
+            = current_linearization_point.block(introspection.block_indices.pressure);
+
+
+        // -------------------------------------------------------------
+        // COMPUTE PARAMETER UPDATES
+        // -------------------------------------------------------------
+
+        // calculate the sensitivity kernels and use them to update the initial guess of density and viscosity
+        // through simple steepest descent. Note: this means the initial density and viscosity is overwritten
+        // for the visualization
+        compute_parameter_update();
+
+        // postprocess unless this is the last iteration because then it will be postprocessed outside of this solver scheme
+        if (i<max_nonlinear_iterations-1)
+          postprocess ();
```

Comment:
What do you do if the solver converges before max_nonlinear_iterations? Then you will postprocess the last iteration twice. Could you instead use the same approach as the other solver schemes, or does this not work for adjoint? `parameters.run_postprocessors_on_nonlinear_iterations` disables the postprocessing after the nonlinear solver finishes, so every nonlinear iterations is only postprocessed once.

```suggestion
        if (parameters.run_postprocessors_on_nonlinear_iterations)
          postprocess ();
```

Reply:
That works really well!